pkg/controller/util.go

```diff
@@ -33,6 +33,37 @@ func (c *Controller) pvcFromKey(key interface{}) (*v1.PersistentVolumeClaim, err
 	return pvc, nil
 }
 
+// checkIfShouldQueuePVC will check our existing pvc looking for
+// specific annotations AnnEndPoint and AnnImportPod
+// AnnEndPoint indicates that the pvc is targeted for our queue
+// AnnImportPod indicates that a pvc is already being processed.
+// Notice due to timing issues, we run this check multiple times.
+func (c *Controller) checkIfShouldQueuePVC(pvc *v1.PersistentVolumeClaim, method string) bool {
+	if metav1.HasAnnotation(pvc.ObjectMeta, AnnEndpoint) {
+		if !metav1.HasAnnotation(pvc.ObjectMeta, AnnImportPod) {
+			// secondary check, just in case, fetching latest pvc object if available to help mitigate race and timing issues
+			// with latency between the store and work queue to double check if we are already processing.
+			glog.Infof("%s: checking latest version of Get pvc for inprocess annotation", method)
+			latest, err := c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+			if err == nil {
+				// checking if we are already processing this request based on latest known Get()
+				if metav1.HasAnnotation(latest.ObjectMeta, AnnImportPod) {
+					glog.Infof("processNextItem: latest Get on pvc reveals that annotation %q exists indicating this was already processed once, skipping pvc\n", method, AnnImportPod)
+					return false
+				}
+			}
+		} else {
+			glog.Infof("%s: pvc has annotation %q, indicating this is being processed or was already processed once, skipping pvc\n", method, AnnImportPod)
+			return false
+		}
+	} else {
+		// this pvc does not contain our endpoint annotation.
+		glog.Infof("%s: annotation %q not found, skipping pvc\n", method, AnnEndpoint)
+		return false
+	}
+	return true
```

Comment:
It looks like this `return` is reached if L48 we have err, or L50 no `AnnImportPod`.
What should be returned if the `Get` api call fails?

Reply:
it's reached if we have our EndPoint anno AND also we have the ImportPod anno

Reply:
if we fail on our API call, we keep going, it's a best effort in my opinion

Reply:
My take on a refactor here:
```
func (c *Controller) checkIfShouldQueuePVC(pvc *v1.PersistentVolumeClaim, method string) bool {
 	if !metav1.HasAnnotation(pvc.ObjectMeta, AnnEndpoint) {
             glog.Infof("%s: annotation %q not found, skipping pvc\n", method, AnnEndpoint)
             return false
        }
        if metav1.HasAnnotation(pvc.ObjectMeta, AnnImportPod) {
             glog.Infof("%s: pvc has annotation %q, indicating this is being processed or was already processed once, skipping pvc\n", method, AnnImportPod)
             return false
        }
 	// secondary check, just in case, fetching latest pvc object if available to help mitigate race and timing issues
 	// with latency between the store and work queue to double check if we are already processing.
 	glog.Infof("%s: checking latest version of Get pvc for inprocess annotation", method)
 	latest, err := c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
 	if err != nil {
            glog.Errorf('unexpected....')
            return false
        }
 	// checking if we are already processing this request based on latest known Get()
 	if metav1.HasAnnotation(latest.ObjectMeta, AnnImportPod) {
 		glog.Infof("processNextItem: latest Get on pvc reveals that annotation %q exists indicating this was already processed once, skipping pvc\n", method, AnnImportPod)
 		return false
 	}
 	return true
}
```

Reply:
if you error because the api fails, you potentially throw away a valid pvc... no?